File: front-end/src/app/domain/models/game/action/impl/card/DrawCardAction.ts
import {GameAction} from "../../GameAction";
import {Card} from "../../../card/Card";
import {Player} from "../../../Player";
import {CardUtil} from "../../../card/CardUtil";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class DrawCardAction extends GameAction {
  private _card: Card;

  constructor(
    id: number,
    actor: Player,
    performedAt: Date,
    card: Card,
  ) {
    super(DrawCardAction.name, id, actor, performedAt);
    this._card = card;
  }

  public static trueCopy(action: DrawCardAction): GameAction | null {
    return new DrawCardAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      CardUtil.trueCopy(action.card),
    );
  }

  public get card() {
    return this._card;
  }

  public override toJson(): object {
    return Object.assign(super.toJson(), {
      "card": this.card.toJson()
    });
  }

  public toString(): string {
    return `${this.actor.user.username} picked ${this.card.toString()}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/card/DrawCardAction.java
package nl.hva.backend.domain.models.game.action.impl.card;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;
import nl.hva.backend.domain.models.game.card.Card;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

/**
 * TODO: Add JavaDoc
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class DrawCardAction extends GameAction {

    @ManyToOne
    private Card card;

    public DrawCardAction(
            Player actor,
            Card card
    ) {
        super(actor);
        this.card = card;
    }

}